src/sgx/libpoet_enclave/poet_enclave.cpp

```diff
@@ -0,0 +1,690 @@
+            inOriginatorPublicKeyHash,
+            inPoetPublicKey,
+            &expectedReportData);
+
+        // Create the enclave report for the caller.
+        sgx_status_t ret =
+            sgx_create_report(
+                inTargetInfo,
+                &expectedReportData,
+                outEnclaveReport);
+        sp::ThrowSgxError(ret, "Failed to create enclave report");
+    } catch (sp::PoetError& e) {
+        Log(
+            POET_LOG_ERROR,
+            "Error in poet enclave(ecall_VerifySignupInfo): %04X -- %s",
+            e.error_code(),
+            e.what());
+        ocall_SetErrorMessage(e.what());
+        result = e.error_code();
+    } catch (...) {
+        Log(
+            POET_LOG_ERROR,
+            "Unknown error in poet enclave(ecall_VerifySignupInfo)");
+        result = POET_ERR_UNKNOWN;
+    }
+
+    return result;
+} // ecall_VerifySignupInfo
+
+uint64_t deserializeWaitCert(const char* waitCert) {
+    JsonValue parsed(json_parse_string(waitCert));
+    sp::ThrowIf<sp::ValueError>(!parsed.value, "Failed to parse SerializedWaitCertificate");
+    JSON_Object* pObject = json_value_get_object(parsed);
+    sp::ThrowIfNull(
+            pObject,
+            "Got Null JSON Object");
+
+    return json_object_dotget_number(pObject, "block_number");
+}
+
+void create_duration(uint8_t* duration, size_t inDurationLenBytes) {
+    sgx_status_t ret = sgx_read_rand(duration, inDurationLenBytes);
+    sp::ThrowSgxError(ret, "Failed to generate duration for wait certificate");
+}
+
+poet_err_t ecall_InitializeWaitCertificate(
+    const char* inPreviousWaitCertificate, 
+    size_t inPreviousWaitCertificateLen, 
+    const char* inValidatorId, 
+    size_t inValidatorIdLen,
+    const char* inPrevBlockId,
+    size_t inPrevBlockIdLen,
+    uint8_t* outDuration,
+    size_t inDurationLenBytes
+    ) 
+{
+    poet_err_t result = POET_SUCCESS;
+    try{
+       
+        sp::ThrowIfNull(
+            inPreviousWaitCertificate,
+            "Previous wait certificate is NULL. It can be empty but not NULL");
+
+        sp::ThrowIfNull(
+            inValidatorId,
+            "Validator ID is NULL");
+
+        sp::ThrowIfNull(
+            inPrevBlockId,
+            "Previous block ID is NULL. It can be empty but not NULL");
+
+        sp::ThrowIfNull(
+            outDuration,
+            "outDuration is NULL");
+
+        sp::ThrowIf<sp::ValueError>(!gValidatorSignupData.initialized, "PPK not created. Cannot initialize wait certificate");
+
+        sp::ThrowIf<sp::ValueError>((inDurationLenBytes != DURATION_LENGTH_BYTES), "Expected 32 bytes duration length");
+
+        create_duration(outDuration, inDurationLenBytes);
+
+        //Check if its first block. And validate for null terminated string and 
+        // the length of string is matching with inPreviousWaitCertificateLen
+        if( strnlen(inPreviousWaitCertificate, inPreviousWaitCertificateLen + 1) != inPreviousWaitCertificateLen ) {
```

Comment:
Since the waitcertificate length is constant. inPreviousWaitCertificateLen is same as the const value before using for check here. The exception here is for genesis block where waitcertificate len will be 0.